File: medium/insert_interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> new_int;
        int i = 0;

        while (i < intervals.size() && newInterval[0] > intervals[i][1]) {
            new_int.push_back(intervals[i]);
            i++;
        }

        while (i < intervals.size() && newInterval[1] >= intervals[i][0]) {
            newInterval[0] = min(newInterval[0], intervals[i][0]);
            newInterval[1] = max(newInterval[1], intervals[i][1]);
            i++;
        }
        new_int.push_back({newInterval[0], newInterval[1]});

        while (i < intervals.size()) {
            new_int.push_back(intervals[i]);
            i++;
        }

        return new_int;
    }
};

File: medium/insert_interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if len(intervals) == 0:
            return [newInterval]

        i = 0

        newArr = []

        newStart = newInterval[0]
        newEnd = newInterval[1]
        
        # Insert normally, no merge
        while i < len(intervals) and intervals[i][1] < newStart:
            newArr.append(intervals[i])
            i += 1

        # Overlapping times, need to merge
        while i < len(intervals) and newEnd >= intervals[i][0]:
            newStart = min(newStart, intervals[i][0])
            newEnd = max(newEnd, intervals[i][1])
            i += 1
        
        newArr.append([newStart, newEnd])

        # No merge needed anymore
        while i < len(intervals):
            newArr.append(intervals[i])
            i += 1

        return newArr



        